tensorflow/compiler/mlir/xla/transforms/buffer_assignment.cc

```diff
@@ -0,0 +1,499 @@
+  /// their potential aliases into account.
+  BufferAssignmentAliasAnalysis aliases;
+};
+
+//===----------------------------------------------------------------------===//
+// BufferAssignment
+//===----------------------------------------------------------------------===//
+
+/// The actual buffer assignment pass that moves alloc and dealloc nodes into
+/// the right positions. It uses the algorithm described at the top of the file.
+// TODO(dfki): create a templated version that allows to match dialect-specific
+// alloc/dealloc nodes and to insert dialect-specific dealloc node.
+struct BufferAssignmentPass : mlir::FunctionPass<BufferAssignmentPass> {
+  void runOnFunction() override {
+    // Get required analysis information first.
+    auto& analysis = getAnalysis<BufferAssignmentAnalysis>();
+
+    // Compute an initial placement of all nodes.
+    llvm::SmallDenseMap<Value, BufferAssignmentPositions, 16> placements;
+    getFunction().walk([&](AllocOp alloc) {
+      placements[alloc] = analysis.computeAllocAndDeallocPositions(alloc);
+    });
+
+    // Move alloc (and dealloc - if any) nodes into the right places
+    // and insert dealloc nodes if necessary.
+    getFunction().walk([&](AllocOp alloc) {
+      // Find already associated dealloc nodes.
+      auto deallocs = analysis.findAssociatedDeallocs(alloc);
+      assert(deallocs.size() < 2 &
+             "Not supported number of associated dealloc operations");
+
+      // Move alloc node to the right place.
+      BufferAssignmentPositions& positions = placements[alloc];
+      Operation* allocOperation = alloc.getOperation();
+      allocOperation->moveBefore(positions.getAllocPosition());
+
+      // If there is an existing dealloc, move it to the right place.
+      if (deallocs.size()) {
+        Operation* nextOp = positions.getDeallocPosition()->getNextNode();
+        if (!nextOp)
+          nextOp = &positions.getDeallocPosition()->getBlock()->back();
+        (*deallocs.begin())->moveBefore(nextOp);
+      } else {
+        // If there is no dealloc node, insert one in the right place.
+        OpBuilder builder(alloc);
+        builder.setInsertionPointAfter(positions.getDeallocPosition());
+        builder.create<DeallocOp>(allocOperation->getLoc(), alloc);
+      }
+    });
+  };
+};
+
+std::unique_ptr<OpPassBase<FuncOp>> createBufferAssignmentPass() {
+  return absl::make_unique<BufferAssignmentPass>();
+}
+
+static PassRegistration<BufferAssignmentPass> buffer_assignment_pass(
+    "buffer-assignment",
+    "Executes buffer assignment pass to automatically move alloc and dealloc "
+    "operations into their proper positions");
+
+/// A simple pass to print debug/test information for the buffer assignment
+/// analysis.
+struct BufferAssignmentTestPass : mlir::FunctionPass<BufferAssignmentTestPass> {
+  void runOnFunction() override {
+    llvm::errs() << "Testing : " << getFunction().getName() << "\n";
```

Comment:
Leftover from debugging?

Reply:
Not actually. This is the approach we used for liveness analysis testing as well. Furthermore, it's used in the test cases to match the functions.

Reply:
Ah, I see. I got confused by the `llvm::errs()`. Why is this printing to `stderr` and not `stdout`?

Reply:
Done